framework/build.gradle

```diff
@@ -78,7 +78,67 @@ checkCompilerMessages {
     ]
 }
 
+task testWPIStubs(type: Test) {
+    description 'Internal task to run the whole-program-inference tests with -Ainfer=stubs to generate stub files'
+    doFirst {
+        delete("tests/whole-program-inference/annotated")
+    }
+    outputs.upToDateWhen { false }
+    include '**/WholeProgramInferenceStubsTest.class'
+    testLogging {
+        // Always run the tests
+        outputs.upToDateWhen { false }
+
+        // Show the found unexpected diagnostics and expected diagnostics not found.
+        exceptionFormat "full"
+        events "passed", "skipped", "failed"
+    }
+}
+
+task testWPIStubsValidate(type: Test) {
+    description 'Internal task to re-run the whole-program-inference tests using the stub files generated by testWPIStubs'
+    outputs.upToDateWhen { false }
+    include '**/WholeProgramInferenceStubsValidationTest.class'
+    testLogging {
+        // Always run the tests
+        outputs.upToDateWhen { false }
+
+        // Show the found unexpected diagnostics and expected diagnostics not found.
+        exceptionFormat "full"
+        events "passed", "skipped", "failed"
+    }
+}
+
+// This task is similar to the wholeProgramInferenceTests task below, but it doesn't
```

Comment:
Another difference is using stub files vs. jaif files.

Would it be possible to re-use logic to reduce cut-and-paste and reduce the likelihood of getting out of sync in the future?  Or to abstract out the doLast block that is identical between the two tasks?